sixtyfps_runtime/corelib/model.rs

```diff
@@ -205,14 +205,51 @@ impl Model for bool {
     }
 }
 
+/// A shared model
+#[repr(transparent)]
+#[derive(derive_more::Deref, derive_more::DerefMut)]
+pub struct SharedModel<T>(pub Rc<dyn Model<Data = T>>);
+
+impl<T> std::fmt::Debug for SharedModel<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "SharedModel(dyn Model)")
+    }
+}
+
+impl<T> Clone for SharedModel<T> {
+    fn clone(&self) -> Self {
+        Self(self.0.clone())
+    }
+}
+
+impl<T> core::cmp::PartialEq for SharedModel<T> {
+    #[allow(clippy::vtable_address_comparisons)]
+    // We think vtable_address_comparisons are safe here!
+    // Rc::ptr_eq will compare a fat pointer consisting of a `vtable` and a `self` pointer.
+    //
+    // We can get:
+    // * false positives: Two different `Model`s get considered the same, even though they are not.
+    //   For this the `self` pointer need to point to the same place. That can only happen if
+    //   `Self` is `()`, which does not make sense for a `Model`. So we ignore this case.
```

Comment:
Self could very well be `()`.

We could imagine for example a  `impl Model for ()` that always returns 0 elements.

or a  `struct MyStaticModel;  impl Model for MyStaticModel { ... }`.

But anyway, this cannot happen that two different model are considered the same, because the Rc adds an allocation for the refcounts and always allocates. so the data pointer will always be different anyway.

Reply:
@ogoffart: You are way deeper in this rabbit hole than me, so did I understand this correctly?
 
 But I think you are right, this lint might actually be wrong: I went ahead and filed https://github.com/rust-lang/rust-clippy/issues/7461 with my current understanding. I am pretty sure I missed something, but maybe some improvement to the explanation of that clippy lint will fall out of this:-)

Reply:
I updated the text, I hope it is a bit better now.